module: bootstrap module loaders in shadow realm

legendecas opened the pull request:
This bootstraps ESM loaders in the ShadowRealm with `ShadowRealm.prototype.importValue` as its entry point and enables loading ESM and CJS modules in the ShadowRealm. The module is imported without a parent URL and resolved with the current process's working directory.

ljharb commented on lib/internal/bootstrap/realm.js:
```diff
@@ -269,6 +269,12 @@ class BuiltinModule {
+  static setRealmAllowRequireByUsers(ids) {
+    canBeRequiredByUsersList = new SafeSet(ids.filter((id) => publicBuiltinIds.includes(id)));
+    canBeRequiredByUsersWithoutSchemeList =
+      new SafeSet(ids.filter((id) => !schemelessBlockList.has(id)));
```
```suggestion
      new SafeSet(ArrayPrototypeFilter(ids, (id) => !schemelessBlockList.has(id)));
```

aduh95 replied:
I think this is actually OK because it's run before any user code, right? If so, we should add comment.

ljharb replied:
Maybe, but it's inside a static function so it *could* be ran multiple times, and realms could potentially be created during runtime?

legendecas replied:
It runs before the user codes. But yes, it is a static method that can run at a later time. I've applied the suggested change. Thanks for the suggestion!

ljharb commented on lib/internal/vm/module.js:
```diff
@@ -120,25 +118,29 @@ class Module {
                                    options.cachedData);
-      const { setCallbackForWrap } = require('internal/modules/esm/utils');
-      setCallbackForWrap(this[kWrap], {
+      registry = {
```
```suggestion
      registry = {
        __proto__: null,
```

legendecas commented:
Rebased on the tip of the main branch and removed dependencies on https://github.com/nodejs/node/pull/48510 (with known limitation on memory leaks). Marking as ready to be reviewed.

joyeecheung commented on lib/internal/bootstrap/shadow_realm.js:
```diff
@@ -0,0 +1,47 @@
+
+// The handler for `ShadowRealm.prototype.importValue`.
+setDefaultImportModuleDynamically((specifier, assertions) => {
+  const { esmLoader } = require('internal/process/esm_loader');
```
Can we emit a warning here about the leaks? (even though shadow realm is already behind flags, I guess an additional warning about something we already know wouldn't hurt...)

mcollina commented:
> Only allowed built-in modules can be imported in the ShadowRealm.

mcollina reviewed (approved):
lgtm

GeoffreyBooth commented:
We’re working on eventually removing the CommonJS loader, leaving only the ESM loader as the single module loader (as it can load both CommonJS and ESM code). Is it possible to refactor this so that `ShadowRealm` always loads everything only through the ESM loader? Then this part of things won’t need refactoring when the CommonJS loader goes away, and there would also be more uniformity in how modules are loaded within `ShadowRealm`.

legendecas commented:
@GeoffreyBooth The ShadowRealm interface only exposes ESM `ShadowRealm.prototype.importValue(speficier)` and no CJS equivalent. The test case in the PR merely shows that the modules imported in the ShadowRealm _can_ create a `require` function, just like what we can do in the principal realm.

Are you suggesting this PR didn't load modules in `ShadowRealm.prototype.importValue(speficier)` with the ESM loader? If so, could you share a hint of how I can fix that?

GeoffreyBooth commented:
> Are you suggesting this PR didn’t load modules in `ShadowRealm.prototype.importValue(speficier)` with the ESM loader? If so, could you share a hint of how I can fix that?

I was just responding to the PR description, “This bootstraps ESM and CJS module loaders in the ShadowRealm.” Did what you mean to say that it allows for loading ESM and CJS _modules_ in the ShadowRealm?

I don’t really follow what a lot of the code in this PR does. If its only internal interactions with module loaders are through the ESM loader, then we’re good.

legendecas commented:
Yes, CJS support in ShadowRealm is enabled by ESM loaders. It seems that the PR description is misleading somehow, I've fixed the wording.

GeoffreyBooth commented:
> The module is imported without a parent URL and resolved with the current process’s working directory.

We had a similar situation with `node:module` `register`, of needing to resolve relative to `process.cwd` as an unfortunate default that we dislike but couldn’t find a better alternative for. We added a second parameter to `register` allowing the user to pass `parentURL` to resolve against, so that most usage would be like `register('ts-node', import.meta.url)`. You might want to consider the same here.

Also please make sure that the string is interpreted as a file URL and _not_ as a path, because `import` statements must be file URLs and you’re calling this API `importValue`. Then users can do things like `realm.importValue(import.meta.resolve('./file.js'), 'foo')`. It should also accept URL instances, to permit `realm.importValue(new URL('./file.js', import.meta.url), 'foo')`.

GeoffreyBooth commented on test/parallel/test-shadow-realm-import-value-resolve.js:
```diff
@@ -0,0 +1,26 @@
+
+async function main() {
+  const realm = new ShadowRealm();
+  const mod = fixtures.path('es-module-shadow-realm', 'state-counter.mjs');
+
+  const dirname = __dirname;
+  // Set process cwd to the parent directory of __dirname.
+  const cwd = path.dirname(dirname);
+  process.chdir(cwd);
+  const relativePath = `./${path.relative(cwd, mod)}`;
+
+  // Make sure that the module can not be resolved relative to __filename.
+  assert.throws(() => require.resolve(relativePath), { code: 'MODULE_NOT_FOUND' });
+
+  // Resolve relative to the current working directory.
+  const getCounter = await realm.importValue(relativePath, 'getCounter');
```
Per the spec, `importValue`‘s first argument is treated the same as `import()`: https://github.com/tc39/proposal-shadowrealm/blob/main/explainer.md#api-typescript-format. That means that this `relativePath` needs to be a `file://` URL string, not a path string, as you can’t `import()` a path string. We have a helper `fixtures.fileURL` that works the same as `fixtures.path` for this, or you can convert like `const relativeFileURL = pathToFileURL(relativePath)`

The examples throughout https://github.com/tc39/proposal-shadowrealm/blob/main/explainer.md use relative URLs like `importValue("./sandbox.js", "test")`. Within Node this is impossible; only methods on `import.meta` know the context of the calling module: https://github.com/nodejs/node/pull/49070. If you could somehow make it possible for `importValue`, other than via the hacky method in https://github.com/nodejs/node/pull/49070, that would be fantastic (and something we’d like to copy for `register`, if possible). As a core language feature like `import()`, I can see it making logical sense that `importValue` might have the same magic “knows the context of the calling module” ability that `import()` has.

ljharb replied:
That's only because `import()` is syntax, and not a global function; `realm.importValue` can't possibly have that magic context because it's API.

legendecas replied:
> That means that this relativePath needs to be a file:// URL string, not a path string, as you can’t import() a path string
> 
> The examples throughout https://github.com/tc39/proposal-shadowrealm/blob/main/explainer.md use relative URLs like `importValue("./sandbox.js", "test")`. Within Node this is impossible

I'm not sure if I'm following the point correctly. AFAICT, the following script is working with Node.js v20.x:

```js
// mod-a.mjs
const mod = await import('./mod-b.mjs');
```

Would you mind clarifying why the first argument of `importValue` should be a URL instead of a relative path string?

GeoffreyBooth replied:
A string like `'./mod-b.mjs'` is ambiguous: it could be either a path string or an URL string. Make it absolute and then it becomes distinct:

- Path string: value like `/Users/you/Sites/app/mod-b.mjs`
- URL string: value like `file:///Users/you/Sites/app/mod-b.mjs`

Specs always want URL strings, since those exist everywhere while path strings can’t exist in browsers. Within Node, you can `import` URL strings but not path strings.

This particular test is written using path strings: `fixtures.path` returns a path string, and all the subsequent operations are path-based rather than URL based. The final value of `relativePath` _might_ be ambiguous, potentially either a path or URL string, but the preceding lines strongly suggest that the author of the test was intending to test that `importValue` accepts a _relative path string,_ not a relative URL string. This is incorrect: `importValue` should only accept URL strings, either relative or absolute.

We should add some more tests to make this crystal clear:

- A test that `importValue` accepts an absolute URL string (such as `file:///Users/you/Sites/app/mod-b.mjs`)
- A test that `importValue` throws when given an absolute path string (such as `/Users/you/Sites/app/mod-b.mjs`)

aduh95 replied:
`./mod-b.mjs` is a relative URL in this context, not a relative path. this only matters if you are not on Windows (because Windows relative path would look like `.\mod-b.mjs`) and the string does not contain any URL special characters (`import './file.cjs?someQueryParam'` would load `file.cjs`, while `require('./file.cjs?someQueryParam')` would load `./file.cjs?someQueryParam.js`; `import './file%20name.js'` would load `./file name.js`, while require('./file%20name.js')` would load `file%20name.js`). So while it's probably fine for most JS devs to ignore the difference between relative URLs and relative paths, that's not something we can do in a Node.js API.

legendecas replied:
Thank you! I believe I understand your concerns better now. I've updated the test with file URL strings instead.

legendecas commented:
@GeoffreyBooth @aduh95 would you mind taking a look at this again? Thank you!

aduh95 commented on lib/internal/bootstrap/shadow_realm.js:
```diff
@@ -0,0 +1,47 @@
+// The handler for `ShadowRealm.prototype.importValue`.
+setDefaultImportModuleDynamically((specifier, assertions) => {
+  const { esmLoader } = require('internal/process/esm_loader');
+  // `parentUrl` is not set in the case of a ShadowRealm top-level import.
```
```suggestion
  // `parentURL` is not set in the case of a ShadowRealm top-level import.
```

aduh95 commented on test/parallel/test-shadow-realm-allowed-builtin-modules.js:
```diff
@@ -0,0 +1,37 @@
+  const mod = fixtures.fileURL('es-module-shadow-realm', 'builtin-modules.mjs');
+  const builtinModulesStr = await realm.importValue(mod, 'builtinModulesStr');
+  const shadowRealmBuiltinModules = builtinModulesStr.split(',');
+  const diff = [];
+  for (const mod of builtinModules) {
+    if (!shadowRealmBuiltinModules.includes(mod)) {
+      diff.push(mod);
+    }
+  }
```
nit
```suggestion
  const diff = builtinModules.filter(mod => !shadowRealmBuiltinModules.includes(mod));
```

aduh95 commented on test/parallel/test-shadow-realm-allowed-builtin-modules.js:
```diff
@@ -0,0 +1,37 @@
+  // Verify that dynamic import can not import disallowed builtin modules.
+  assert.ok(!shadowRealmBuiltinModules.includes('fs'));
+  await assert.rejects(realm.importValue('fs', 'readFileSync'), {
+    message: /Cannot find package 'fs'/,
```
Should we check the error code instead?

aduh95 replied:
Do you want to add a comment explaining that?

legendecas replied:
Yeah, thank you for the suggestion!

aduh95 commented on test/parallel/test-shadow-realm-allowed-builtin-modules.js:
```diff
@@ -0,0 +1,37 @@
+  assert.throws(() => {
+    innerRequireValue('fs', 'readFileSync');
+  }, {
+    message: /Cannot find module 'fs'/,
```
ditto

aduh95 replied:
```suggestion
    // As above, we can only validate the error message, not the error code.
    message: /Cannot find module 'fs'/,
```

joyeecheung commented on src/module_wrap.cc:
```diff
@@ -579,31 +590,43 @@ static MaybeLocal<Promise> ImportModuleDynamically(
-    auto it = env->id_to_function_map.find(id);
-    CHECK_NE(it, env->id_to_function_map.end());
-    object = it->second->object();
+  if (options->Length() == HostDefinedOptions::kLength) {
```
What are we getting as host defined options for import coming from shadow realms? Is it just empty fixed array? I think it would be more natural to tweak this branch as `if (realm->kind() == Realm::Kind::kPrincipal)`..

legendecas replied:
> I think it would be more natural to tweak this branch as `if (realm->kind() == Realm::Kind::kPrincipal)`

Only the import initiated by `ShadowRealm.prototype.importValue` has the empty host-defined option. The referrer script is still present for `import` statement and dynamic import so the host-defined option would not be empty. In that case, this branch applies in ShadowRealm too.

joyeecheung commented on src/module_wrap.cc:
```diff
@@ -627,13 +650,13 @@ static MaybeLocal<Promise> ImportModuleDynamically(
   CHECK(args[0]->IsFunction());
   Local<Function> import_callback = args[0].As<Function>();
-  env->set_host_import_module_dynamically_callback(import_callback);
+  realm->set_host_import_module_dynamically_callback(import_callback);
```
Instead of mixing a default callback into code paths that aren't exercised by shadow realms, can we just configure a different handler for the realm using this binding in `bootstrap/shadow_realm.js`? I think the `isolate->SetHostImportModuleDynamicallyCallback(ImportModuleDynamically);` call below can probably just moved to `node::NewIsolate()` (it might be a good idea to follow up with a IsolateSetting option for embedders to opt out of this, though that's out of scope of this PR).

legendecas replied:
Dynamic import can be initiated in a ShadowRealm, since `ShadowRealm.prototype.importValue` can import a module that imports another module. So this callback is still needed for ShadowRealm to have proper module support.

joyeecheung replied:
What I mean is we just set this callback to the specialized function in `bootstrap/realm.js`, instead of routing it as a default case in the handler used by the principal realm.

legendecas replied:
Thanks, this is really a good approach! Updated.

joyeecheung commented on src/node_process.h:
```diff
@@ -83,7 +89,7 @@ class BindingData : public SnapshotableObject {
   static constexpr size_t kBufferSize =
       std::max(sizeof(uint64_t), sizeof(uint32_t) * 3);
-  v8::Global<v8::ArrayBuffer> array_buffer_;
+  bool HrTimeBufferValid();
```
Can this be split into a different PR/commit? It would be easier to read to split the `env` -> `realm`/`isolate_data` changes into another commit too, it's a bit difficult to hunt for the important changes from all the refactorings ;)

legendecas replied:
Thanks for the suggestion, I've split the PR into smaller commits.

joyeecheung commented on src/node_shadow_realm.cc:
```diff
@@ -32,6 +40,10 @@ MaybeLocal<Context> HostCreateShadowRealmContextCallback(
+
+  // We do not expect the realm bootstrapping to throw any
+  // exceptions. If it does, exit the current Node.js instance.
+  TryCatchScope try_catch(env, TryCatchScope::CatchMode::kFatal);
```
Note that `TryCatchScope` uses enhancers from the main context...as we are refactoring a bunch of env access to realms already, it might be safer to refactor that and especially `ReportFatalException`.

legendecas replied:
This `TryCatchScope` here is only used when the built-in modules throw errors in development. Yet it is true that it should be refactored to enhance the error in a specific realm since it calls into JavaScript. I'll do this in a follow-up PR.

joyeecheung replied:
okay, I am okay with a follow-up, just seeing that there are already a bunch of env -> realm refactorings, wondered if you want to just do it together here.

joyeecheung commented on src/node_shadow_realm.cc:
```diff
@@ -137,6 +149,25 @@ v8::MaybeLocal<v8::Value> ShadowRealm::BootstrapRealm() {
   }
 
+  if (ExecuteBootstrapper(
+          "internal/bootstrap/switches/does_not_own_process_state")
```
What's the idea about choosing which script to run and which not to run? This just sets up a bunch of states in process, but there are also a bunch of other states initialized in `bootstrap/node.js` too. I wonder if we should move those to `realm.js` and just leave the `globalThis` mutations there..

legendecas replied:
This is needed to set up the process properties for built-in modules exclusively since `process` is not exposed to ShadowRealm yet. I added a code comment to explain that `process.cwd()` initialized in this switch is needed by the ESM loader.

I believe a bunch of global properties like `Buffer` can be moved to `realm.js` indeed. I can do it in a follow-up work.

joyeecheung replied:
I am okay with a follow-up though want to clarify that I was talking about the "adding things to `process`"  part in `node.js`, not the "exposing `process` to `globalThis` part - the former could just be moved to realm.js, if we are already loading `does_not_own_process_state` here which is also just "adding things to `process`".

legendecas commented:
@joyeecheung would you mind taking a look at this again? thank you!

joyeecheung commented on lib/internal/bootstrap/shadow_realm.js:
```diff
@@ -0,0 +1,39 @@
+} = require('internal/bootstrap/realm');
+
+BuiltinModule.setRealmAllowRequireByUsers([
+  'assert',
```
What is the criteria of allowing modules in this list? It seems somewhat arbitrary to me if `module` is allowed but `fs` isn't (but you can monkey patch `module` to read files and `process.dlopen()` arbitrary things?)

joyeecheung commented:
Also I see that there is a proposal of the demotion of shadow realms to stage 2 in the upcoming TC39 meeting https://github.com/tc39/agendas/blob/main/2023/09.md It seems the primary reason is:

> SpiderMonkey’s position is that the proposal is not implementable without:
> - Review of which DOM APIs are exposed to ShadowRealms
> - Sufficient tests to ensure the correct behaviour of these APIs in ShadowRealms

I wonder if we also need some proper review about which APIs/modules are supposed to be exposed to ShadowRealms. Of course it's behind a flag right now but I imagine if the idea about what's allowed in ShadowRealms and what's not on the web is changed in general is changed, we need to make sure our criteria is in sync with the browsers or at least not too far from it..

legendecas commented:
@joyeecheung thanks for the suggestion. I agree that the preliminary built-in module list can be revisited and focus on `importValue` support in this PR. I have updated the PR accordingly.

legendecas commented on lib/internal/modules/esm/utils.js:
```diff
@@ -93,16 +93,37 @@ async function importModuleDynamicallyCallback(wrap, specifier, assertions) {
+  }
+
+  // The handler for `ShadowRealm.prototype.importValue`.
+  const { esmLoader } = require('internal/process/esm_loader');
```
I've added tests to verify that both `--require` and `--loader` is not enabled in the ShadowRealm. However, I'm still open to enabling the custom loader in the ShadowRealm when it is verified it doesn't manipulate objects across realms. So I'd distinguish the loader worker's case from the shadow realm's.

GeoffreyBooth replied:
FYI we're moving away from `--loader` toward `--import`. Might deprecate/remove `--loader`, but not for a bit. Don't know if this affects this PR but I thought I'd mention.

legendecas commented:
Rebased to resolve conflicts.

joyeecheung commented on lib/internal/modules/esm/utils.js:
```diff
@@ -165,20 +165,41 @@ async function importModuleDynamicallyCallback(symbol, specifier, assertions) {
  */
-function initializeESM(isLoaderWorker = false) {
+function initializeESM(isLoaderWorker = false, isShadowRealm = false) {
   _isLoaderWorker = isLoaderWorker;
```
I think we can just update this to `_forceDefaultLoader` and rename the argument so it can be used for shadow realms too. The `isShadowRealm` is essentially doing the same thing as `isLoaderWorker`, there's no need to invent a new way to bypass the custom loader.

joyeecheung commented on src/module_wrap.cc:
```diff
@@ -577,11 +595,23 @@ static MaybeLocal<Promise> ImportModuleDynamically(
+    ProcessEmitWarning(
+        env,
+        "Module support in ShadowRealm is experimental and known to leak "
+        "memory.");
```
This can probably be removed now.

joyeecheung commented on test/parallel/test-shadow-realm-preload-module.js:
```diff
@@ -0,0 +1,23 @@
+
+async function main() {
+  // Verifies that --require preload modules are not enabled in the ShadowRealm.
+  const child = await common.spawnPromisified(process.execPath, [
```
Can you use `spawnSyncAndExitWithoutError` from `test/common/child_process` here instead? That would dump more information to the logs if the process doesn't exit in the way expected for easier debugging for the CI.

joyeecheung commented on src/module_wrap.cc:
```diff
@@ -559,14 +564,27 @@ static MaybeLocal<Promise> ImportModuleDynamically(
+  // When the dynamic import is initiated by a realm top-level request, e.g.
+  // ShadowRealm.prototype.importValue, the host defined options should be
+  // empty.
+  bool is_realm_top_level_import =
```
Notice that the bug reported in https://github.com/nodejs/node/issues/49726 can also trigger this even when it does not come from a shadow realm.

joyeecheung replied:
Actually I wonder if we can extend the HostImportModuleDynamically callback to explicitly pass this information in the arguments...the length alone doesn't seem to be a very good reference here.

joyeecheung replied:
Actually, I wonder why we don't get a referrer in this case...[the spec](https://tc39.es/proposal-shadowrealm/#sec-shadowrealmimportvalue) says:

> 6. Perform [HostImportModuleDynamically](https://tc39.es/ecma262/#sec-hostimportmoduledynamically)(null, specifierString, innerCapability).

I wonder why the referrer needs to be null here instead of e.g. the `[[ShadowRealm]]` slot of the shadow realm object. Is is by design that the host gets no information about the referrer here? That seems a bit odd..

legendecas replied:
This is different from the eval case.

`ShadowRealm.prototype.importValue` is a built-in API and should avoid resolving the module specifiers by determining who called it, which would be a case of dynamic scope that we should avoid.

As for propagating active script in V8, I would suggest waiting for discussion at https://github.com/tc39/ecma262/pull/3195.

legendecas replied:
> There is currently an edge case in V8 such that some code that makes no use of shadow realms can result in an empty referrer
> Which means the edge case now would fail DCHECK_EQ(realm->kind(), Realm::kShadowRealm) which doesn't seem quite right...

I believe this case would not be able to reach the DCHECK because there is an `is_realm_top_level_import && realm->kind() != Realm::kShadowRealm` branch to disallow empty host-defined options in non-shadow-realm realms.

However, this is true for `Promise.resolve('import(`./mod`)').then(eval)` in ShadowRealm, which would not throw a `TypeError: Invalid host defined options` but the module resolution would proceed. Indeed, this can be an inconsistency between the ShadowRealms and principal realms.

I'll submit a separate PR to add a fallback for empty host-defined options in all realms.

mcollina reviewed (approved):
lgtm

legendecas commented:
Seems like `parallel/test-runner-output` is failing across different jobs because negative durations were reported. My first attempt to reproduce the failure locally didn't work out. I'll try to figure out why it is failing.

legendecas commented:
@joyeecheung would you mind taking a look at this PR again? Thank you!

joyeecheung reviewed (commented):
Mostly LGTM with a few comments

legendecas commented on lib/internal/process/per_thread.js:
```diff
@@ -65,10 +64,10 @@ function refreshHrtimeBuffer() {
   // the upper/lower 32 bits of the second part of the value,
   // and the remaining nanoseconds of the value.
-  hrValues = new Uint32Array(binding.hrtimeBuffer);
+  hrValues = binding.hrtimeBuffer;
```
Fixed in https://github.com/nodejs/node/pull/48655/commits/f6a1dff4b8384b9c4e718e2545f0df555c5256ac.

legendecas commented on lib/internal/process/esm_loader.js:
```diff
@@ -10,6 +10,9 @@ const { kEmptyObject, getCWDURL } = require('internal/util');
+    esmLoader = createModuleLoader(useCustomLoadersIfPresent);
+  },
   get esmLoader() {
     return esmLoader ??= createModuleLoader(true);
```
Fixed in https://github.com/nodejs/node/pull/48655/commits/1520e25edc662c8d919daf1adec3fe9f394f85fb, thanks for pointing this out!

legendecas commented:
Landed in ce4a06d5dc18...fc2862b7f55b
